From 7df4faf7578cd36174ced5b019be28fb5ab1e322 Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Mon, 24 Sep 2012 08:44:56 -0700
Subject: Fixed segfault caused by use-after-free

When removing a configuration from the supplicant, the wpa_config
struct will be freed in the process. In Android, a notification
will be sent to user space for this event. However, the message
reads the SSID, which has already been freed. There are two places
pointing to the same SSID struct, and this patch explicitly set
one to NULL after removal of configuration. This fixes the segfault
that may appear when "forgetting" a network in user space.

Issue: AXIA-538
Change-Id: Ia1d62d7c0a9f76f258d0dbba01254c45e07d345d
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 wpa_supplicant/ctrl_iface.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/wpa_supplicant/ctrl_iface.c b/wpa_supplicant/ctrl_iface.c
index ad0392f..4302670 100644
--- a/wpa_supplicant/ctrl_iface.c
+++ b/wpa_supplicant/ctrl_iface.c
@@ -2295,6 +2295,16 @@ static int wpa_supplicant_ctrl_iface_remove_network(
 		wpa_sm_set_config(wpa_s->wpa, NULL);
 		eapol_sm_notify_config(wpa_s->eapol, NULL, NULL);
 
+#ifdef ANDROID
+		/*
+		 * Android will send a notification with ssid,
+		 * but the ssid struct has already been freed.
+		 * Set it explicitly to NULL so there is no
+		 * segfault.
+		 */
+		wpa_s->current_ssid = NULL;
+#endif
+
 		wpa_supplicant_deauthenticate(wpa_s,
 					      WLAN_REASON_DEAUTH_LEAVING);
 	}
-- 
1.7.4.1

